File: Radix-sort/python/simple.py
import random

arr = [random.randint(0,120) for _ in range(6)]
# arr = [34, 119, 42, 59, 64, 56]
print(arr)

def sort(arr):
    max = arr[0]
    for v in arr:
        if v > max:
            max = v

    totalBits = len(str(max))
    sorted = [0]*len(arr)
    digit = 1
    for _ in range(totalBits):
        bucket = [0]*10
        for v in arr:
            bucketIdx = int(v/digit)%10
            bucket[bucketIdx] += 1
        for i in range(1,10):
            bucket[i] += bucket[i-1]
        for j in range(len(arr)-1,-1,-1):
            bucketIdx = int(arr[j]/digit)%10
            idx = bucket[bucketIdx] - 1
            sorted[idx] = arr[j]
            bucket[bucketIdx] -= 1
        arr = [v for v in sorted]
        digit *=10

    return arr


print(sort(arr))

File: Radix-sort/js/simple.js
let arr = []

function initData(count){
    for(let i = 0 ; i < count ; i++){
        arr.push(Math.floor(Math.random()*150))
    }
}

initData(10)
// arr = [ 78, 121, 6, 118, 0, 36 ]
console.log(arr)

function sort(){
    let max = arr[0]
    for(let i = 1 ; i < arr.length ; i++){
        if(arr[i] > max){
            max = arr[i]
        }
    }

    let totalPos = (max+"").length
    // console.log(totalPos)
    const bucket = new Array(10)

    for(let i = 1 , digit = 1 ; i <= totalPos ; i++, digit *= 10){
        for(let m = 0 ; m < arr.length ; m++){
            let bucketIdx = Math.floor(arr[m]/digit)%10
            // console.log("bucketIdx:",bucketIdx,bucket[bucketIdx])
            if(!bucket[bucketIdx]){
                bucket[bucketIdx] = [arr[m]]
            }else{
                bucket[bucketIdx].push(arr[m])
            }
        }
        // console.log("bucket:",bucket)
        let idx = 0
        for(let n = 0 ; n < 10 ; n++){
            while(bucket[n] && bucket[n].length > 0){
                arr[idx] = bucket[n].shift()
                idx++
            }
        }
        console.log(arr)
    }
}

sort()